core/execution.go

```diff
@@ -75,6 +79,101 @@ func Create(env vm.Environment, caller vm.ContractRef, code []byte, gas, gasPric
 	return ret, address, err
 }
 
+// Create2 creates a new contract with the given code
+func Create2(env vm.Environment, caller vm.ContractRef, code []byte, gas, gasPrice, salt, value *big.Int) (ret []byte, address common.Address, err error) {
+	addr := crypto.CreateAddress2(caller.Address(), common.BigToHash(salt).Bytes(), crypto.Keccak256(code))
+	ret, address, err = create(env, caller, &addr, nil, crypto.Keccak256Hash(code), nil, code, gas, gasPrice, value, false)
+	// ret, address, err = exec(env, caller, &addr, nil, crypto.Keccak256Hash(code), nil, code, gas, gasPrice, value, false)
+	// Here we get an error if we run into maximum stack depth,
+	// See: https://github.com/ethereum/yellowpaper/pull/131
+	// and YP definitions for CREATE
+
+	//if there's an error we return nothing
+	if err != nil && err != vm.ErrRevert {
+		return nil, address, err
+	}
+	return ret, address, err
+}
+
+// create creates a new contract using code as deployment code.
+func create(env vm.Environment, caller vm.ContractRef, address, codeAddr *common.Address, codeHash common.Hash, input, code []byte, gas, gasPrice, value *big.Int, readOnly bool) ([]byte, common.Address, error) {
+	evm := env.Vm()
+	// Depth check execution. Fail if we're trying to execute above the
+	// limit.
+	if env.Depth() > callCreateDepthMax {
+		caller.ReturnGas(gas, gasPrice)
+
+		return nil, common.Address{}, errCallCreateDepth
+	}
+	if !env.CanTransfer(caller.Address(), value) {
+		caller.ReturnGas(gas, gasPrice)
+
+		return nil, common.Address{}, ErrInsufficientFunds
+	}
+	nonce := env.Db().GetNonce(caller.Address())
+	env.Db().SetNonce(caller.Address(), nonce+1)
+
+	// Ensure there's no existing contract already at the designated address
+	// contractHash := evm.StateDB.GetCodeHash(address)
+	contractHash := env.Db().GetCodeHash(*address)
+	if env.Db().GetNonce(*address) != 0 || env.Db().GetCode(*address) != nil || (contractHash != (common.Hash{}) && contractHash != crypto.Keccak256Hash(nil)) {
+		return nil, common.Address{}, errContractAddressCollision
```

Comment:
Isn't this related to EIP 684? If so, probably shouldn't be implemented unless approved by community